pkg/api/compatibility_test.go

```diff
@@ -0,0 +1,207 @@
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util/fielderrors"
+)
+
+func TestCompatibility_v1_ServiceAccount(t *testing.T) {
+	// Test "spec.host" -> "spec.nodeName"
+	expectedHost := "my-host"
+	// Test "spec.serviceAccount" -> "spec.serviceAccountName"
+	expectedServiceAccount := "my-service-account"
+	// Test "tcp" protocol gets converted to "TCP" and validated
+	originalProtocol := "tcp"
+	expectedProtocol := "TCP"
+
+	input := []byte(fmt.Sprintf(`
+{
+	"kind":"Pod",
+	"apiVersion":"v1",
+	"metadata":{"name":"my-pod-name", "namespace":"my-pod-namespace"},
+	"spec": {
+		"host":"%s",
+		"serviceAccount":"%s",
+		"containers":[{
+			"name":"my-container-name",
+			"image":"my-container-image",
+			"ports":[{"containerPort":1,"protocol":"%s"}]
+		}]
+	}
+}
+`, expectedHost, expectedServiceAccount, originalProtocol))
+
+	testCompatibility(t, "v1", input,
+		func(obj runtime.Object) fielderrors.ValidationErrorList {
+			return validation.ValidatePod(obj.(*api.Pod))
+		},
+		map[string]string{
+			"spec.host":     expectedHost,
+			"spec.nodeName": expectedHost,
+
+			"spec.serviceAccount":     expectedServiceAccount,
+			"spec.serviceAccountName": expectedServiceAccount,
+
+			"spec.containers[0].ports[0].protocol": expectedProtocol,
+		},
+	)
+}
+
+func TestCompatibility_v1_Service(t *testing.T) {
+	// Test "spec.portalIP" -> "spec.clusterIP"
+	expectedIP := "1.2.3.4"
+	// Test "tcp" protocol gets converted to "TCP" and validated
+	originalProtocol := "tcp"
+	expectedProtocol := "TCP"
+
+	input := []byte(fmt.Sprintf(`
+{
+	"kind":"Service",
+	"apiVersion":"v1",
+	"metadata":{"name":"my-service-name", "namespace":"my-service-namespace"},
+	"spec": {
+		"portalIP":"%s",
+		"ports":[{"port":1,"protocol":"%s"}]
+	}
+}
+`, expectedIP, originalProtocol))
+
+	testCompatibility(
+		t,
+		"v1",
+		input,
+		func(obj runtime.Object) fielderrors.ValidationErrorList {
+			return validation.ValidateService(obj.(*api.Service))
+		},
+		map[string]string{
+			"spec.portalIP":          expectedIP,
+			"spec.clusterIP":         expectedIP,
+			"spec.ports[0].protocol": expectedProtocol,
+		},
+	)
+}
+
+func TestCompatibility_v1_Endpoints(t *testing.T) {
+	// Test "tcp" protocol gets converted to "TCP" and validated
+	originalProtocol := "tcp"
+	expectedProtocol := "TCP"
+
+	input := []byte(fmt.Sprintf(`
+{
+	"kind":"Endpoints",
+	"apiVersion":"v1",
+	"metadata":{"name":"my-endpoints-name", "namespace":"my-endpoints-namespace"},
+	"subsets": [{
+		"addresses":[{"ip":"1.2.3.4"}],
+		"ports": [{"port":1,"targetPort":1,"protocol":"%s"}]
+	}]
+}
+`, originalProtocol))
+
+	testCompatibility(
+		t,
+		"v1",
+		input,
+		func(obj runtime.Object) fielderrors.ValidationErrorList {
+			return validation.ValidateEndpoints(obj.(*api.Endpoints))
+		},
+		map[string]string{
+			"subsets[0].ports[0].protocol": expectedProtocol,
+		},
+	)
+}
+
+func testCompatibility(
```

Comment:
added t.Log() messages pointing to this PR... the messages get surfaced if the test fails